src/Dropdown.js

```diff
@@ -88,82 +95,130 @@ const defaultProps = {
  */
-class Dropdown extends React.Component {
-  static displayName = 'ReactOverlaysDropdown';
-
-  static getDerivedStateFromProps({ drop, alignEnd, show }, prevState) {
-    const lastShow = prevState.context.show;
-    return {
-      lastShow,
-      context: {
-        ...prevState.context,
-        drop,
-        show,
-        alignEnd,
-      },
-    };
+function Dropdown({
+  drop,
+  alignEnd,
+  defaultShow,
+  show: rawShow,
+  onToggle: rawOnToggle,
+  itemSelector,
+  focusFirstItemOnShow,
+  children,
+}) {
+  const forceUpdate = useForceUpdate();
+  const { show, onToggle } = useUncontrolled(
+    { defaultShow, show: rawShow, onToggle: rawOnToggle },
+    { show: 'onToggle' },
+  );
+
+  const [toggleElement, setToggle] = useCallbackRef();
+
+  // We use normal refs instead of useCallbackRef in order to populate the
+  // the value as quickly as possible, otherwise the effect to focus the element
+  // may run before the state value is set
+  const menuRef = useRef();
+  const menuElement = menuRef.current;
+
+  const setMenu = useCallback(
+    ref => {
+      menuRef.current = ref;
+      // ensure that a menu set triggers an update for consumers
+      forceUpdate();
+    },
+    [forceUpdate],
+  );
+
+  const lastShow = usePrevious(show);
+  const lastSourceEvent = useRef(null);
+  const focusInDropdown = useRef(false);
+
+  const toggle = useCallback(
+    event => {
+      onToggle(!show, event);
+    },
+    [onToggle, show],
+  );
+
+  const context = useMemo(
+    () => ({
+      toggle,
+      drop,
+      show,
+      alignEnd,
+      menuElement,
+      toggleElement,
+      setMenu,
+      setToggle,
+    }),
+    [
+      toggle,
+      drop,
+      show,
+      alignEnd,
+      menuElement,
+      toggleElement,
+      setMenu,
+      setToggle,
+    ],
+  );
+
+  if (menuElement && lastShow && !show) {
+    focusInDropdown.current = menuElement.contains(document.activeElement);
   }
 
-  constructor(...args) {
-    super(...args);
-
-    this._focusInDropdown = false;
-
-    this.menu = null;
-
-    this.state = {
-      context: {
-        close: this.handleClose,
-        toggle: this.handleClick,
-        menuRef: r => {
-          this.menu = r;
-        },
-        toggleRef: r => {
-          const toggleNode = r && ReactDOM.findDOMNode(r);
-          this.setState(({ context }) => ({
-            context: { ...context, toggleNode },
-          }));
-        },
-      },
-    };
-  }
+  const focusToggle = useEventCallback(() => {
+    if (toggleElement && toggleElement.focus) {
+      toggleElement.focus();
+    }
+  });
 
-  componentDidUpdate(prevProps) {
-    const { show } = this.props;
-    const prevOpen = prevProps.show;
+  const maybeFocusFirst = useEventCallback(() => {
+    const type = lastSourceEvent.current;
+    let focusStype = focusFirstItemOnShow;
 
-    if (show && !prevOpen) {
-      this.maybeFocusFirst();
+    if (focusStype == null) {
```

Comment:
Does it make sense to `useMemo` this outside of this callback? I don't think this needs to be calculated on-the-fly.

Reply:
in practice it only runs on show so it's probably fine